indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/DefaultIndexTaskInputRowIteratorBuilder.java

```diff
@@ -0,0 +1,146 @@
+ *      {@link GranularitySpec} bucket intervals, invoke the absent bucket interval {@link Consumer}
+ *      callback.
+ *
+ * If any of the handlers invoke their respective callback, the {@link HandlingInputRowIterator} will yield
+ * a null {@link InputRow} next; otherwise, the next {@link InputRow} is yielded.
+ * </pre>
+ */
+public class DefaultIndexTaskInputRowIteratorBuilder implements IndexTaskInputRowIteratorBuilder
+{
+  private CloseableIterator<InputRow> delegate = null;
+  private GranularitySpec granularitySpec = null;
+  private HandlingInputRowIterator.InputRowHandler nullRowHandler = null;
+  private HandlingInputRowIterator.InputRowHandler absentBucketIntervalHandler = null;
+  private List<HandlingInputRowIterator.InputRowHandler> appendedInputRowHandlers = new ArrayList<>();
+
+  @Override
+  public DefaultIndexTaskInputRowIteratorBuilder delegate(CloseableIterator<InputRow> inputRowIterator)
+  {
+    this.delegate = inputRowIterator;
+    return this;
+  }
+
+  @Override
+  public DefaultIndexTaskInputRowIteratorBuilder granularitySpec(GranularitySpec granularitySpec)
+  {
+    this.granularitySpec = granularitySpec;
+    return this;
+  }
+
+  @Override
+  public DefaultIndexTaskInputRowIteratorBuilder nullRowRunnable(Runnable nullRowRunnable)
+  {
+    this.nullRowHandler = inputRow -> {
+      if (inputRow == null) {
+        nullRowRunnable.run();
+        return true;
+      }
+      return false;
+    };
+    return this;
+  }
+
+  @Override
+  public DefaultIndexTaskInputRowIteratorBuilder absentBucketIntervalConsumer(
+      Consumer<InputRow> absentBucketIntervalConsumer
+  )
+  {
+    this.absentBucketIntervalHandler = inputRow -> {
+      Optional<Interval> intervalOpt = granularitySpec.bucketInterval(inputRow.getTimestamp());
+      if (!intervalOpt.isPresent()) {
+        absentBucketIntervalConsumer.accept(inputRow);
+        return true;
+      }
+      return false;
+    };
+    return this;
+  }
+
+  @Override
+  public HandlingInputRowIterator build()
+  {
+    Preconditions.checkNotNull(delegate, "delegate required");
+    Preconditions.checkNotNull(granularitySpec, "granularitySpec required");
+    Preconditions.checkNotNull(nullRowHandler, "nullRowRunnable required");
+    Preconditions.checkNotNull(absentBucketIntervalHandler, "absentBucketIntervalConsumer required");
+
+    ImmutableList.Builder<HandlingInputRowIterator.InputRowHandler> handlersBuilder = ImmutableList.<HandlingInputRowIterator.InputRowHandler>builder()
+        .add(nullRowHandler)
+        .add(createInvalidTimestampHandler())
+        .add(absentBucketIntervalHandler)
+        .addAll(appendedInputRowHandlers);
+
+    return new HandlingInputRowIterator(delegate, handlersBuilder.build());
+  }
+
+  /**
+   * @param inputRowHandler Optionally, append this input row handler to the required ones.
+   */
+  DefaultIndexTaskInputRowIteratorBuilder appendInputRowHandler(HandlingInputRowIterator.InputRowHandler inputRowHandler)
```

Comment:
nit: `@VisibleForTesting`

Or indicate that we can add additional row handlers in the top level javadoc

Reply:
Added to javadoc.